### PR TITLE
LLM Layer testcases

### PR DESCRIPTION
**Pull request by PhaneeshB**
- Add dummy layers test for most llama layers 
- Add real weights test for rms_norm + lm_head layers

### CODE REVIEW

File: sharktank/sharktank/utils/_helpers.py
Comment (rsuderman) [inline sharktank/sharktank/utils/_helpers.py:1]: Lets try to focus in on a reusable component rather than a monlithic end-to-end-runner. We want a thing that takes a `torch.Module` and generates an equivalent module that uses iree as a backend. This thing should do the following:

- Export the MLIR functions from each
- Compile the exported mlir
- Load the module into a runtime with device setup
- Return the invokeable module

We don't want to include executing it as once exported it could be invoked multiple times to test.

As a simple V0, lets assume a single `forward` function. Export the module. And reload into an iree runtime.

Comment (PhaneeshB) [reply sharktank/sharktank/utils/_helpers.py:1]: Pushed changes with refactor of the existing method to:
1. Export and capture eager outputs
2. Compile
3. Run with IREE to generate output
4. compare eager vs iree outputs

File: sharktank/sharktank/utils/_helpers.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    # NOTE: assuming args are not nested.
+    empty_args = tuple([
+        torch.empty(arg.shape, dtype=arg.dtype) for arg in args
+    ])
+
+    # need to get this info from the test, currently only for static shapes
+    # one corresponding to each arg
+    dynamic_shapes = tuple([dict() for _ in args])
+
+    print(f"dyn shapes : {dynamic_shapes}")
```
[the remark this answers is not in the document]
Comment (PhaneeshB) [reply sharktank/sharktank/utils/_helpers.py:85]: That's a debug print that got in.
will take it out, since the output values for real layers generally to big to see as a printed tensor.

File: sharktank/sharktank/utils/_helpers.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    load_iree_module,
+    prepare_iree_module_function_args,
+    run_iree_module_function,
+    iree_to_torch,
+)
+
+from iree.turbine.aot import *
+# from iree.turbine.aot import FxProgramsBuilder, export
+
+DEFAULT_COMPILE_FLAGS = [
```
[the remark this answers is not in the document]
Comment (PhaneeshB) [reply sharktank/sharktank/utils/_helpers.py:20]: yup this is to get things going
thinking of a config setup in layers test to have model specific flags as the next step
another way would be a per-test basis flag-list to be added as a pytest option but, that would make running the test messy.

File: sharktank/tests/layers/linear_with_iree_test.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        super().__init__()
+        self.lin = torch.nn.Linear(in_f, out_f, bias=bias, dtype=dtype)
+
+    def forward(self, x):
+        return self.lin(x)
+
+@pytest.mark.parametrize("dtype,atol", [(torch.float32, 1e-4), (torch.float16, 1e-4)])
+def test_linear_iree_vs_eager(dtype, atol):
+    torch.manual_seed(0)
+    m = Linear(64, 64, bias=False, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (PhaneeshB) [reply sharktank/tests/layers/linear_with_iree_test.py:23]: that might not be what we are testing here
for layers based tests we are comparing torch vs iree execution for different sharktank llm model layers

The goal for the layer based test suite is to have layers that correspond to actual llm model layers across variants (for llama) that we export in sharktank.

In order to get started, we added tests that have a close but dummy implementation ( same structure, smaller size, random weights) 
current linear layer test is one of these dummy tests.

in the next steps we want to update this to create linear layer using an irpa file 
when adding that test we shall keep this as a mock test to facilitate running the test without an irpa file.

File: sharktank/sharktank/utils/_helpers.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        name=target_fn,
+        args=empty_args,
+        dynamic_shapes=(dynamic_shapes,),
+        strict=False, 
+        )
+    def _(module, *fn_args):
+        return module.forward(*fn_args)
+
+    export_output = export(fxb, import_symbolic_shape_expressions=True)
+    export_output.save_mlir(mlir_path)
```
[the remark this answers is not in the document]
Comment (PhaneeshB) [reply sharktank/sharktank/utils/_helpers.py:98]: For the current usecase we are saving this in a tmp dir 
from a stand alone method pov, we just expect it to export the IR along with outputs

File: sharktank/tests/layers/output_lm_test_with_iree.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        self.hp = config.hp
+        
+        # Output normalization layer
+        self.output_norm = RMSNormLayer(
+            theta("output_norm"), 
+            epsilon=self.hp.attention_layer_norm_rms_epsilon
+        )
+        
+        # Output linear layer (language model head)
+        self.output_lm_head = LinearLayer(
```
[the remark this answers is not in the document]
Comment (PhaneeshB) [reply sharktank/tests/layers/output_lm_test_with_iree.py:32]: yup this is a composite test since RMS Norm + Linear is the last step in getting the logits output
This serves as a starting point for adding other tests

File: sharktank/sharktank/utils/iree.py [hunk 6]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+
+def _as_tuple(x):
+    if isinstance(x, tuple):
+        return x
+    if isinstance(x, list):
+        return tuple(x)
+    return (x,)
+
+
+def export_torch_module_to_mlir(
```
Comment (sogartar) [inline sharktank/sharktank/utils/iree.py]: The place of this function is more appropriate in `utils/export.py`. It seems that you are trying to make another overload of the underlying of IREE Turbine's [export](https://github.com/iree-org/iree-turbine/blob/c80577fdc61a1565cd906205f82c8c79c62f5aea/iree/turbine/aot/exporter.py#L171) that can export to a file instead of returning an `ExportOutput` object. I suggest that you use the same arguments as the underlying export function with the addition of `mlir_path`. The we don't need deal with a slightly different naming of arguments like `target_fn` instead of `function_name`.

File: sharktank/sharktank/utils/iree.py [hunk 7]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+    expected = torch_output
+    if isinstance(expected, torch.Tensor):
+        expected = (expected,)
+
+    actual = ()
+    for idx, o in enumerate(iree_output):
+        if isinstance(o, np.ndarray):
+            output = torch.tensor(o).type_as(expected[idx])
+        else:
+            output = o
```
Comment (sogartar) [inline sharktank/sharktank/utils/iree.py]: I would expect that this would use the existing conversion functions for IREE device array to torch tensor before doing the comparison.

File: sharktank/sharktank/utils/iree.py [hunk 8]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+
+    return with_iree_device_context(run_with_devices, devices)
+
+
+def compare_iree_torch_outputs(
+    iree_output,
+    torch_output,
+    *,
+    atol=1e-4,
+    rtol=0.0,
```
Comment (sogartar) [inline sharktank/sharktank/utils/iree.py]: These defaults are not appropriate at this level. This is supposed to be a general function. If it has any defaults they should most likely be whatever the underlying `torch.testing.assert_close` does.

File: sharktank/sharktank/utils/iree.py [hunk 9]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+            device=devices[0],
+            function_name=entrypoint,
+        )
+        results = iree_to_torch(*iree_out)
+        return results
+
+    return with_iree_device_context(run_with_devices, devices)
+
+
+def compare_iree_torch_outputs(
```
Comment (sogartar) [inline sharktank/sharktank/utils/iree.py]: A more appropriate name is `assert_iree_torch_outputs_close`.

File: sharktank/sharktank/utils/iree.py [hunk 10]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+        )
+        # Compare outputs
+        compare_iree_torch_outputs(
+            iree_output=iree_output,
+            torch_output=torch_output,
+            atol=atol,
+            rtol=rtol,
+        )
+    del iree_devices
+    gc.collect()
```
Comment (sogartar) [inline sharktank/sharktank/utils/iree.py]: What kind of problems did you run into that requires manual GC?
Is it something related to the explanation in [with_iree_device_context](https://github.com/nod-ai/shark-ai/blob/d5c56eb0007799f39581413025fa5bc0f63fd707/sharktank/sharktank/utils/iree.py#L223).

File: sharktank/tests/layers/rotary_embedding_hf_test.py [hunk 11]
```diff
@@ -524,10 +526,12 @@ def run_iree_module(iree_devices):
                 device=iree_devices[0],
                 function_name="rotary_openweight_fw",
             )
-
-            return iree_to_torch(*iree_result)
+            iree_result_torch = iree_to_torch(*iree_result)
+            iree_result_torch = tuple(t.clone() for t in iree_result_torch)
```
[the remark this answers is not in the document]
Comment (monorimet) [reply sharktank/tests/layers/rotary_embedding_hf_test.py]: I think reworking `iree_to_torch` to take a `to_host` bool arg and hiding `clone().detach()` behind that might be the best way -- this was indeed redundant, in original comment's PR state

File: sharktank/sharktank/utils/iree.py [hunk 12]
```diff
@@ -753,3 +759,251 @@ def run_model_with_iree_run_module(
...
+        skip("No IRPA path provided. Use --parameters to specify the IRPA file.")
+
+    # Skip test if IRPA file doesn't exist
+    if not Path(irpa_path).exists():
+        skip(f"IRPA file not found: {irpa_path}")
+
+    return irpa_path
+
+
+def run_iree_vs_torch_fx(
```
[the remark this answers is not in the document]
Comment (monorimet) [reply sharktank/sharktank/utils/iree.py]: changing fx to eager
